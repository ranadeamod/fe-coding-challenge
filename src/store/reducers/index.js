import { combineReducers } from "redux";
import {SELECT_CELL, DECLARE_WINNER, RESTART} from "../actions/moves";

export const createBoard = (i) =>
  Array(i)
    .fill(null)
    .map(_ =>
      Array(i)
        .fill(null)
    )

const boardInitialState = createBoard(3);
const gameInitialState = { currentPlayer: 'X', winner: null };

export const board = (state = boardInitialState, action) => {
  switch (action.type) {
    case SELECT_CELL: {
      const newBoard = JSON.parse(JSON.stringify(state))
      newBoard[action.row][action.col] = action.currentPlayer
      return newBoard
    }
    case RESTART: {
      return boardInitialState;
    }
    default: {
      return state
    }
  }
}

export const game = (state = gameInitialState, action) => {
  switch (action.type) {
    case SELECT_CELL: {
      return {
        ...state,
        currentPlayer: state.currentPlayer === 'X' ? 'O' : 'X'
      }
    }
    case DECLARE_WINNER: {
      return {
        ...state,
        winner: action.winner
      };
    }
    case RESTART: {
      return gameInitialState;
    }
    default: {
      return state
    }
  }
}

export default combineReducers({
  board,
  game
})
