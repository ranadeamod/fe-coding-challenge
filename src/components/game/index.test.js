import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from '../../store';
import { createBoard } from '../../store/reducers';
import { Game } from '.';

describe('<Game />', () => {
    it('matches the snapshot', () => {
        const props = {
            board: createBoard(3),
            player: null
        }
        const tree = renderer.create(<Provider store={configureStore()}><Game {...props} /></Provider>);
        expect(tree).toMatchSnapshot();
    });
})
