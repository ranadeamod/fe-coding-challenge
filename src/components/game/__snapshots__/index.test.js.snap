// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> matches the snapshot 1`] = `
<div
  className="Game-Grid"
>
  <div
    className="Game-Grid-Row"
  >
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
  </div>
  <div
    className="Game-Grid-Row"
  >
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
  </div>
  <div
    className="Game-Grid-Row"
  >
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
    <div
      className="Game-Grid-Cell"
      onClick={[Function]}
    >
      <span
        className="unselected"
      >
        ?
      </span>
    </div>
  </div>
</div>
`;
